File: Searching/JavaScript/depthFirstSearch.js
class Graph {
    constructor() {
        this.graph = {};
    }

    addEdge(starting_vector, ending_vector) {
        if(this.graph[starting_vector] == undefined) {
            this.graph[starting_vector] = [ending_vector];
        } else {
            this.graph[starting_vector].push(ending_vector);
        }
    }

    dfsUtil(vector, visited) {
        visited[vector] = true;
        console.log(vector);

        for(let i = 0; i < this.graph[vector].length; i++) {
            if(!visited[this.graph[vector][i]]) {
                this.dfsUtil(this.graph[vector][i], visited);
            }
        }
    }

    dfs(starting_vector) {
        let visited = {};
        this.dfsUtil(starting_vector, visited);
    }
}

graph = new Graph();
graph.addEdge(0, 1);
graph.addEdge(0, 2);
graph.addEdge(1, 2);
graph.addEdge(2, 0);
graph.addEdge(2, 3);
graph.addEdge(3, 3);
graph.dfs(2);

File: Searching/Ruby/depth_first_search.rb
class Graph 
    def initialize 
        @graph = Hash.new { Array.new }
    end

    def add_edge(starting_vector,  ending_vector)
        @graph[starting_vector] += [ending_vector]
    end 

    def dfs(starting_vector)
        visited = Hash.new { false }
        dfs_util(starting_vector, visited)
    end 

    private 

    def dfs_util(vector, visited)
        visited[vector] = true
        puts vector

        @graph[vector].each do |v|
            dfs_util(v, visited) unless visited[v]
        end
    end
end

graph = Graph.new

graph.add_edge(0, 1)
graph.add_edge(0, 2)
graph.add_edge(1, 2)
graph.add_edge(2, 0)
graph.add_edge(2, 3)
graph.add_edge(3, 3)

graph.dfs(2)